File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.cpp
#include <string>
#include <vector>

using namespace std;

vector<vector<int>> edgeList;
vector<int> visited;
int maxSheep = 0;

void dfs(vector<int>& info, int sheep, int wolf, vector<int> possible) {
    maxSheep = max(maxSheep, sheep);
    for (int i = 0; i < possible.size(); i++) {
        vector<int> tmp = possible;
        int nsheep = sheep, nwolf = wolf;
        int p = tmp[i];
        
        if (info[p] == 0) { // sheep
            nsheep++;
        } else { // wolf
            nwolf++;
            if (nwolf >= nsheep) continue;
        }
        tmp.erase(tmp.begin() + i);
        for (auto& e : edgeList[p]) tmp.push_back(e);
        visited[p] = true;
        dfs(info, nsheep, nwolf, tmp);
        visited[p] = false;
    }   
}

int solution(vector<int> info, vector<vector<int>> edges) {
    int n = info.size();
    edgeList.assign(n, {});
    visited.assign(n, false);
    for (auto& e : edges) {
        edgeList[e[0]].push_back(e[1]);
    }
    visited[0] = true;
    dfs(info, 1, 0, edgeList[0]);
    return maxSheep;
}

File: 프로그래머스/3/92343. 양과 늑대/양과 늑대.java
import java.util.*;

class Solution {
    List<Integer>[] edgeList;
    int answer = 0;
    int[] info;
    int n;
    
    public int solution(int[] info, int[][] edges) {
        this.info = info;
        n = info.length;
        edgeList = new ArrayList[n];
        
        for (int i = 0; i < n; i++) {
            edgeList[i] = new ArrayList<>();
        }
        
        for (int[] edge : edges) {
            edgeList[edge[0]].add(edge[1]);
        }
        
        dfs(1);
        return answer;
    }
    
    private void dfs(int mask) {
        int sheep = 0;
        int wolf = 0;
        
        for (int node = 0; node < n; node++) {
            if ((mask & (1 << node)) == 0) continue;
            
            if (info[node] == 0) {
                sheep++;
            } else {
                wolf++;
            }
        }
        
        if (wolf >= sheep) return;
        
        answer = Math.max(answer, sheep);
        
        for (int current = 0; current < n; current++) {
            if ((mask & (1 << current)) == 0) continue;
            
            for (int next : edgeList[current]) {
                if ((mask & (1 << next)) != 0) continue;
                
                dfs(mask | (1 << next));
            }
        }
    }
}